**Problem Statement**

The rise in online transactions has led to an increase in credit card fraud making it a significant concern for financial institutions and cardholders Identifying fraudulent transactions in real-time is crucial for minimizing financial losses and maintaining trust. The challenge lies in efficiently detecting fraudulent behavior while minimizing false positives and ensuring that legitimate transactions are not flagged incorrectly.

In [ ]:
from google.colab import files
uploaded = files.upload()

# **Importing the Dependencies**

**Importing the libaries**

In [54]:
# Importing the necessary libaries

import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

# **Exploratory Data Analysis**

**Reading and understanding the data**

In [55]:
# loading the dataset to a Pandas DataFrame
credit_card_data = pd.read_csv('/content/file.csv')

In [56]:
# print first 5 rows of the dataset
credit_card_data.head()

Time        V1        V2        V3        V4  ...       V26       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  1.378155  ... -0.189115  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  0.448154  ...  0.125895 -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  0.379780  ... -0.139097 -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993 -0.863291  ... -0.221929  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  0.403034  ...  0.502292  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [57]:
# checking the shape(no of rows and columns) of the dataset
credit_card_data.shape

(284807, 31)

In [58]:
# displaying the last 5 rows of the dataset
credit_card_data.tail()

Time         V1         V2        V3  ...       V27       V28  Amount  Class
284802  172786.0 -11.881118  10.071785 -9.834783  ...  0.943651  0.823731    0.77      0
284803  172787.0  -0.732789  -0.055080  2.035030  ...  0.068472 -0.053527   24.79      0
284804  172788.0   1.919565  -0.301254 -3.249640  ...  0.004455 -0.026561   67.88      0
284805  172788.0  -0.240440   0.530483  0.702510  ...  0.108821  0.104533   10.00      0
284806  172792.0  -0.533413  -0.189733  0.703337  ... -0.002415  0.013649  217.00      0

[5 rows x 31 columns]

In [59]:
# dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [60]:
# another method for checking the number of missing values in each column
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

**Checking the distribution of legit transactions and fraudulent transactions**

In [61]:
# distributions of legit transactions (0) and fraudulent transactions (1) in the 'Class' column
credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

In highly imbalanced datasets like this one, where legitimate transactions far outnumber fraudulent ones, comparing the mean values of features for both classes (Class = 0 for legitimate, Class = 1 for fraud) helps us identify significant differences. These differences are crucial for understanding patterns in fraudulent transactions, enabling us to focus on key features that help distinguish between fraud and legitimate behavior. This comparison not only aids in feature selection but also improves model performance by overcoming the imbalance and ensuring the model captures important fraud-related characteristics.

In [62]:
# Separating the dataset into legitimate and fraudulent transactions for analysis

legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]


In [63]:
# Printing the number of rows and columns for both legitimate and fraudulent transactions
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [64]:
# statistical measures of the data

legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [65]:
# display statistical summaries of the Amount column
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [66]:
# Comparing the mean values of all features for both legitimate (Class = 0) and fraudulent (Class = 1) transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3  ...       V26       V27       V28      Amount
Class                                              ...                                          
0      94838.202258  0.008258 -0.006271  0.012171  ... -0.000089 -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778 -7.033281  ...  0.051648  0.170575  0.075667  122.211321

[2 rows x 30 columns]

# **Under-Sampling**

Build a sample dataset containing similar distribution of normal transactions and Fraudulent transactions.

Number of Fraudulent transactions --> 492

In [67]:
# Taking a random sample of legitimate transactions to match the number of fraudulent transactions (492)
# This helps in balancing the dataset for training purposes by creating an equal number of legitimate and fraudulent samples.
legit_sample = legit.sample(n = 492)

Concatenating two DataFrames

In [68]:
# Concatenating the sampled legitimate transactions and all fraudulent transactions along the rows (axis = 0).
# This creates a new balanced dataset with an equal number of legitimate and fraudulent transactions for analysis.
new_dataset = pd.concat([legit_sample, fraud], axis = 0)

In [69]:
# Displaying the first few rows of the newly created balanced dataset to verify the structure and content.
new_dataset.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
166829  118332.0  2.078643  0.184265 -1.717599  ... -0.060291 -0.030129    1.98      0
257054  157981.0  2.024362 -0.293731 -2.084112  ... -0.119150 -0.083987   39.95      0
70945    54083.0  1.055734  0.236773  0.568670  ...  0.035906  0.038097   60.00      0
281151  169969.0 -1.572974  0.829427  1.579634  ...  0.366859 -0.090650  379.29      0
267600  162844.0  1.973858 -0.210757 -2.016852  ... -0.108649 -0.070326   85.80      0

[5 rows x 31 columns]

In [70]:
# Displaying the last few rows of the newly created balanced dataset
new_dataset.tail()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
279863  169142.0 -1.927883  1.125653 -4.518331  ...  0.292680  0.147968  390.00      1
280143  169347.0  1.378559  1.289381 -5.004247  ...  0.389152  0.186637    0.76      1
280149  169351.0 -0.676143  1.126366 -2.213700  ...  0.385107  0.194361   77.89      1
281144  169966.0 -3.113832  0.585864 -5.399730  ...  0.884876 -0.253700  245.00      1
281674  170348.0  1.991976  0.158476 -2.583441  ...  0.002988 -0.015309   42.53      1

[5 rows x 31 columns]

In [71]:
# Checking the distribution of the 'Class' column in the new balanced dataset.
# This ensures that the dataset now contains an equal number of legitimate (Class = 0) and fraudulent (Class = 1) transactions.
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [72]:
# Calculating the mean of all features for each class (legitimate and fraudulent) in the new balanced dataset.
# This helps in comparing feature values between legitimate (Class = 0) and fraudulent (Class = 1) transactions to identify any differences.
new_dataset.groupby('Class').mean()

Time        V1        V2        V3  ...       V26       V27       V28      Amount
Class                                              ...                                          
0      98132.705285  0.094448 -0.001166 -0.008342  ... -0.001721 -0.048865  0.009382   80.003618
1      80746.806911 -4.771948  3.623778 -7.033281  ...  0.051648  0.170575  0.075667  122.211321

[2 rows x 30 columns]

# **Separating the features (X) and the target variable (Y) in the new dataset.**

X contains all the columns except 'Class' (the features), while Y contains the 'Class' column (the labels indicating legitimate or fraudulent transactions).

In [73]:
X = new_dataset.drop(columns = 'Class', axis = 1)
Y = new_dataset['Class']

In [74]:
# Displaying the feature set (X) to verify that it contains all the relevant features for model training, excluding the 'Class' column.
print(X)

            Time        V1        V2        V3  ...       V26       V27       V28  Amount
166829  118332.0  2.078643  0.184265 -1.717599  ...  0.173654 -0.060291 -0.030129    1.98
257054  157981.0  2.024362 -0.293731 -2.084112  ...  0.553323 -0.119150 -0.083987   39.95
70945    54083.0  1.055734  0.236773  0.568670  ... -0.423407  0.035906  0.038097   60.00
281151  169969.0 -1.572974  0.829427  1.579634  ...  0.526028  0.366859 -0.090650  379.29
267600  162844.0  1.973858 -0.210757 -2.016852  ...  0.638074 -0.108649 -0.070326   85.80
...          ...       ...       ...       ...  ...       ...       ...       ...     ...
279863  169142.0 -1.927883  1.125653 -4.518331  ...  0.788395  0.292680  0.147968  390.00
280143  169347.0  1.378559  1.289381 -5.004247  ...  0.739467  0.389152  0.186637    0.76
280149  169351.0 -0.676143  1.126366 -2.213700  ...  0.471111  0.385107  0.194361   77.89
281144  169966.0 -3.113832  0.585864 -5.399730  ...  0.606116  0.884876 -0.253700  245.00
281674  17

In [75]:
# Displaying the target variable (Y) to verify the class labels (legitimate or fraudulent).
print(Y)

166829    0
257054    0
70945     0
281151    0
267600    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


#**Splitting the dataset into training and testing sets.**

The training data (X_train, Y_train) will be used to train the model, and the testing data (X_test, Y_test) will be used to evaluate the model's performance.

In [76]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify = Y, random_state = 2)

In [77]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Model Training

Logistic Regression

In [78]:
model = LogisticRegression(max_iter=1000)

In [79]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
model.fit(X_train_scaled, Y_train)


LogisticRegression(max_iter=1000)

In [85]:
# define the Logistic Regression class
class LogisticRegression():
    def __init__(self, lr, epochs):
        # initialize hyperparameters
        self.lr = lr
        self.epochs = epochs

    def fit(self, X, y):
        # get number of data and features
        self.num_of_data, self.features = X.shape

        # initialize the parameters
        self.W = np.zeros(self.features)
        self.b = 0

        self.X = X
        self.y = y

        for i in range(self.epochs):
            self.update_hyperparam()

    def update_hyperparam(self):
        # calculate y_hat
        y_hat = 1 / (1 + np.exp( - (self.X.dot(self.W) + self.b)))

        # derivative
        dw = (1/self.num_of_data) * np.dot(self.X.T, (y_hat - self.y))
        db = (1/self.num_of_data) * np.sum(y_hat - self.y)

        # update weight and bias
        self.W = self.W - self.lr*dw
        self.b = self.b - self.lr*db

    def predict(self, X):
        y_pred = 1 / (1 + np.exp(- (X.dot(self.W) + self.b)))
        y_pred = np.where(y_pred < 0.5, 0, 1)
        return y_pred

In [86]:
# initailize the standardScaler to standardize the feature set
scaler = StandardScaler()

# fit the scaler to the feature data (X) and transform it to standardized features
scaler.fit(X)
std_features = scaler.transform(X)

# Reassign the standardized features to X and the target variable to y
X = std_features
y = Y

print(X)
print(y)

[[ 0.5991571   0.8001174  -0.44679578 ... -0.11507468 -0.1624335
  -0.4590186 ]
 [ 1.42138398  0.79028558 -0.57805649 ... -0.17098368 -0.28284612
  -0.28319512]
 [-0.73321585  0.61483901 -0.43237676 ... -0.02369856 -0.0098995
  -0.19035181]
 ...
 [ 1.657171    0.30114619 -0.18808889 ...  0.30800088  0.33946627
  -0.10751057]
 [ 1.66992466 -0.14038962 -0.33651414 ...  0.78272071 -0.66227911
   0.66630721]
 [ 1.67784644  0.78441949 -0.45387767 ... -0.05496722 -0.12929998
  -0.2712482 ]]
166829    0
257054    0
70945     0
281151    0
267600    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


#**Model Evaluation & Accuracy Score**

Model evaluation and accuracy scoring are crucial for assessing a model's performance. After training, it's important to check how well the model performs on both training and test data to ensure it generalizes well and isn;t overfitting. This helps determine the model's reliability and effectiveness in making accurate predictions, especially for real-world applications like fraud detection.

In [88]:
# Split the dataset into training and testing sets with 80% for training and 20% for testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 2)

#printing the following details :

print("Features sized")
print(X.shape)
print(X_train.shape)
print(X_test.shape)

print("Target sized")
print(y.shape)
print(y_train.shape)
print(y_test.shape)

Features sized
(984, 30)
(787, 30)
(197, 30)
Target sized
(984,)
(787,)
(197,)


In [89]:
# Set the learning rate and the number of epochs for training the logistic regression model
lr = 0.01
epochs = 50000

# Initialize the custom Logistic Regression model with specified learning rate and epochs
classifier = LogisticRegression(lr = lr, epochs = epochs)

# Train the model using the training data
classifier.fit(X_train,y_train)

# make predictions on the training data
train_prediction = classifier.predict(X_train)

# calculate the accuracy of the model on training data
train_accuracy_score = accuracy_score(y_train, train_prediction)*100

# make predictions on the test data
test_prediction = classifier.predict(X_test)

# calculate the accuracy of the model on the test data
test_accuracy_score = accuracy_score(y_test, test_prediction)*100

print("\n========== My Custom Model Summary ==========\n")
# Display the learning rate and number of epochs used for training
print(f"Learnig Rate: {lr}")
print(f"Epochs: {epochs} \n")
# Display the accuracy on the training and test datasets
print(f"Accuracy on train data: {train_accuracy_score}%")
print(f"Accuracy on test data: {test_accuracy_score}%")


========== My Custom Model Summary ==========

Learnig Rate: 0.01
Epochs: 50000 

Accuracy on train data: 94.91740787801778%
Accuracy on test data: 93.90862944162437%


The model evaluation reveals that the custom logistic regression model has achieved an accuracy of 94.92% on the training data and 93.91% on the test data. These results indicate that the model is performing well, with strong predictive capabilities on both the data it was trained on and unseen test data. The close similarity between the training and test accuracies suggests minimal overfitting, meaning the model is generalizing effectively to new data. Overall, the model evaluation shows promising performance, with high accuracy in detecting patterns, which is crucial for tasks such as fraud detection in imbalanced datasets.